packages/serve/src/startDevServer.ts

```diff
@@ -11,21 +11,25 @@ import Server from 'webpack-dev-server/lib/Server';
  */
 export default function startDevServer(compiler, options): void {
     const firstWpOpt = compiler.compilers ? compiler.compilers[0].options : compiler.options;
-    const devServerOptions = firstWpOpt.devServer || {};
+    let devServerOptions = firstWpOpt.devServer || {};
+
+    // set default host and port
+    devServerOptions = { host: 'localhost', port: 8080, ...devServerOptions };
 
-    const host = options.host || devServerOptions.host || 'localhost';
-    const port = options.port || devServerOptions.port || 8080;
     // socket should not have a default value, because it should only be used if the
     // user explicitly provides it
     const socket = options.socket || devServerOptions.socket;
 
-    options.host = host;
-    options.port = port;
     if (socket) {
         options.socket = socket;
     }
 
-    const server = new Server(compiler, options);
+    // Compose config from the CLI and devServer object from webpack config
+    const serverConfig = { ...devServerOptions, ...options };
```

Comment:
Problem - this way we won't be able to override default options unless specified via flag. eg. `clientLogLevel` by default is set to 'info', so if we pass any other option in devServer, it won't take precedence, but we also need to support flags like `--hot` and override them with the devServer config so that's a dilemma

Reply:
CLI args should be override config values, what is a problem, can you clarify with examples?

Reply:
If we don't need to override config values then I think we're fine with devServer options taking precedence.